runner/host/memory.go

```diff
@@ -1,32 +1,80 @@
 package host
 
 import (
+	"context"
 	"time"
 
 	"github.com/hashicorp/go-hclog"
+	"github.com/shirou/gopsutil/v3/mem"
+
 	"github.com/hashicorp/hcdiag/op"
 	"github.com/hashicorp/hcdiag/runner"
-	"github.com/shirou/gopsutil/v3/mem"
 )
 
 var _ runner.Runner = Memory{}
 
-type Memory struct{}
+type Memory struct {
+	ctx     context.Context
+	Timeout runner.Timeout
+}
 
 func (m Memory) ID() string {
 	return "memory"
 }
 
+func NewMemory(timeout runner.Timeout) *Memory {
+	return NewMemoryWithContext(context.Background(), timeout)
+}
+
+func NewMemoryWithContext(ctx context.Context, timeout runner.Timeout) *Memory {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	return &Memory{
+		ctx:     ctx,
+		Timeout: timeout,
+	}
+}
+
 // Run calls out to mem.VirtualMemory
 func (m Memory) Run() op.Op {
 	startTime := time.Now()
+	resChan := make(chan op.Op, 1)
+
+	var runCtx context.Context
+	var cancel context.CancelFunc
+	if 0 < m.Timeout {
+		runCtx, cancel = context.WithTimeout(m.ctx, time.Duration(m.Timeout))
+		defer cancel()
+	}
+
+	go func(resChan chan op.Op, startTime time.Time) {
+		o := m.run()
+		o.Start = startTime
```

Comment:
I like this! I wonder, though, if the enclosing `startTime` could be used instead of passing in a parameter to this anonymous function. 🤔

Reply:
Shadowing the var name to pass it into the closure is a bad practice, but copying the startTime to avoid modification helps a little bit with system robustness. I shared some implementation ideas out of band for an immutable `Start` type for `Op.Start`, but I'm going to kick that can down the road until a genuine need arises.